src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
@ReubenBond It seems like, and please correct me if I'm mistaken, the concern is regarding how grain activator will custom construct grains, and in particular, how that will work with facets.  Is that correct?

The facet pattern in this PR is a pattern we support but do not enforce.  It does not limit user’s construction of their grain in any way.  The pattern, as a means of exposing Orleans features to grains, should be used to expose features, but the features should not depend on being exposed by attributes on arguments in a constructor, as the feature should not care how it gets wired up, only that it is.  With the introduction of DI, Orleans is extremely extensible, and by exposing (some parts of) the grain lifecycle to the application layer we allow for a much wider range of Orleans extensions.  However, we can't be responsible for supporting every possible way one might want to use this power, nor would it be desirable (even internally) to have several different patterns be used, due to the testability and maintainability problems we're already facing.

> how grain construction is performed and what is acceptable for an IGrainActivator

So the answer, imo, to the grain activator question is, they can use the facet pattern we support using grain constructors and attributes, or they can use the features directly wiring things up themselves.

The facet system design we came up with (and agreed on) was that features would be injected in constructors with metadata provided by attributes.  How is this to be done without constructor selection?  There are other ways we can wire this up.  We can modify the IgrainActivator interface to take arguments and pass the constructed facets to it.  We can construct the facets when creating the IGrainActivationContext and expose them to the grain activator via the context.  I’m sure there are other approaches as well, but all of which require constructor selection because the design requires access to the metadata stored in attributes on the parameters.  Am I missing something?

This PR is the first pass at the facet system.  More work is necessary, and the system will be fleshed out more as we use it.  In particular, the current constructor selection is incorrect, and the test facets do not have a strong separation between the feature logic and the attribution.  If you'd prefer I address these in this PR, I'm willing to do that, as the constructor selection logic will eventually need to be done anyway.  What I'm less willing to do is make the wiring of features to the grain an implementation detail of the facet rather than a common supported pattern.  A central goal of this system was to have a common supported pattern for exposing features to grains.  If there are alternatives with technical advantages, I'm happy to adopt them, but alternatives need to address the requirements and have technical advantages.

Regarding the GrainTypeData. It is serializable so grain type data can be sent to the client and other silos (heterogeneous deployments), but the constructor arguments are only ever required locally, so there is no reason to serialize them.  It already contains logic and data, it’s not simply a raw data type, and it’s where we keep static data about grain types, which is what the constructor parameters (imo) are.  Where the parameters cache is stored is entirely orthogonal to the question of constructor selection, which seems to be the main objection.  Again I note that this is internal logic that can be changed at any time without customer impact should we find a better place for this.  If we want to add some extensibility point for the application layer to provide it's own metadata to facets rather than getting the information from the constructors, making this cache unnecessary, I'm not opposed to that, but blocking this PR and the continued development of these systems on new requirements and advanced features no one has requested is, imo, poor prioritization at best, more likely speculative over engineering.